README.md
# financial-life
A framework for analysing financial products in personalized contexts

<table>
<tr>
	<td>
	Latest Release
	</td>
	<td>
	<img src="https://img.shields.io/pypi/v/FinancialLife.svg" alt="latest release" />
	</td>
</tr>
<table>

[CHANGELOG.md](CHANGELOG.md)

# Description

FinancialLife is an opinionated framework written in Python that allows to simulate monetary flows between different types of accounts. These simulations allow a deeper understanding of financial plans and a better comparison of financial products (in particular loan conditions) for personal circumstances. With FinancialLife you can

* analyse loan conditions and payment strategies
* describe the properties of your financial plans with a few lines of code
* create dynamic monetary flows between accounts for modeling more realistic scenarios
* extend the code by controller functions (e.g. for modeling tax payments)

View [documentation](docs/README.md) for a more detailed introduction.

# Example
Say you want to model an account with regular income and payments to a loan

```python
from FinancialLife.financing import accounts as a
from datetime import timedelta, datetime

# create a private bank account and a loan account
account = a.BankAccount(amount = 1000, interest = 0.001, name = 'Main account')
loan = a.Loan(amount = 100000, interest = 0.01, name = 'House Credit')

# add these accounts to the simulation
simulation = a.Simulation(account, loan)

# describe monetary flows between accounts
simulation.add_regular('Income', account, 2000, interval = 'monthly')
simulation.add_regular(account, loan, lambda: min(1500, -loan.account), interval = 'monthly')

# simulate for ten years
simulation.simulate(delta = timedelta(days=365*10))

# plot the data
simulation.plt_summary()

# print reports summarized by years
print(account.report.yearly())
print(loan.report.yearly())

# analyze data
print("Interests on bank account: %.2f" % sum(account.report.yearly().interest))
print("Interests on loan account: %.2f" % sum(loan.report.yearly().interest))
```  

The output will look like this:

<img src="docs/img/simple_example_01_small.png" alt="Simple simulation in FinancialLife" width="800">


	Main account
	Date          account     output     input    interest
	----------  ---------  ---------  --------  ----------
	31.12.2016    2000.32   -3000.00   4000.00        0.32
	31.12.2017    8005.58  -18000.00  24000.00        5.26
	31.12.2018   14016.85  -18000.00  24000.00       11.27
	31.12.2019   20034.13  -18000.00  24000.00       17.28
	31.12.2020   26057.42  -18000.00  24000.00       23.29
	31.12.2021   32086.74  -18000.00  24000.00       29.32
	31.12.2022   46271.00   -9853.30  24000.00       37.56
	31.12.2023   70330.32       0.00  24000.00       59.32
	31.12.2024   94413.68       0.00  24000.00       83.36
	31.12.2025  118521.15       0.00  24000.00      107.47
	01.10.2026  138521.15       0.00  20000.00        0.00
	House Credit
	Date          account    interest    payment
	----------  ---------  ----------  ---------
	31.12.2016  -97190.22     -190.22    3000.00
	31.12.2017  -80064.23     -874.01   18000.00
	31.12.2018  -62766.98     -702.75   18000.00
	31.12.2019  -45296.76     -529.78   18000.00
	31.12.2020  -27652.02     -355.26   18000.00
	31.12.2021   -9830.65     -178.63   18000.00
	31.12.2022       0.00      -22.65    9853.30
	31.12.2023       0.00        0.00       0.00
	31.12.2024       0.00        0.00       0.00
	31.12.2025       0.00        0.00       0.00
	Interests on bank account: 374.45
	Interests on loan account: -2853.30

Now let's say, we put some money on a special savings account with better interests, because we want to purchase in two years a car. With FinancialLife, you just add the necessary changes to your model.

```python
# create new account
savings = a.BankAccount(amount = 5000, interest = 0.007, name = 'Savings')

# add it to the simulation (or create a new simulation with all three accounts)
simulation.add_account(savings)

# add regular payment to the savings-account
simulation.add_regular(account, savings, 500, interval = 'monthly')

# somewhere in the distant future we will make a payment to
# the vendor of a car
simulation.add_unique(savings, 'Vendor of a car', 10000, '17.03.2019')
```

The plot will now include the savings-account as well.

<img src="docs/img/simple_example_02_small.png" alt="Simple simulation in FinancialLife" width="800">

You can also export the simulation to HTML to explore your model in the browser:

```python
from FinancialLife.reports import html

html.report(simulation, style="standard", output_dir = result_folder)
```

<img src="docs/img/html_summary_01.png" alt="Simple simulation in FinancialLife" width="800" height="407">

You can analyse the reports as [pandas](https://github.com/pandas-dev/pandas) DataFrame as well and export it to excel:

```python
import pandas as pd
from FinancialLife.reports import excel

account.report.as_df()    # Hello pandas
excel.report(simulation, filename='reports.xls')  # explore the results in excel

```

[Here](FinancialLife/examples/README.md) are more examples. FinancialLife supports:
* [dependencies between accounts](FinancialLife/examples/dependencies.py), e.g. to model how the ownership of a property rises when the loan decreases
* [meta-data](FinancialLife/examples/meta_data.md), e.g. for writing tax-calculations, which require additional knowledge about your payments
* [controller-functions](FinancialLife/examples/meta_data.md) for dynamic changes of the simulation properties during simulation

# Installation

FinancialLife is available in version 0.9.2. It is written in Python 3.4 and has not been tested for Python 2.x.

To get a working environment, simply do

	git clone https://github.com/MartinPyka/FinancialLife.git
	cd FinancialLife
	virtualenv venv
	source venv/bin/activate
	pip install -r requirements.txt
	# test an example
	python FinancialLife/examples/simple_examples.py

For installing the package:

	git clone https://github.com/MartinPyka/FinancialLife.git
	cd FinancialLife
	python setup.py install

Or use pip

 	pip install FinancialLife

You can checkout the example with

	python FinancialLife/examples/simple_examples.py

# Why FinancialLife

FinancialLife was designed with the idea in mind that any line of code should contribute to the description of the problem you want to model. In spreadsheets, you would deal with a lot of auxiliary tables to accurately calculate the course of a loan influenced by incoming payments and generated interests. In FinancialLife, you just create your loan account with the given interests rate and you define the regular payments going into this loan account. That's it. Changes in the model and the exploration of different parameters within this model are therefore way easier to accomplish than in a spreadsheet-based simulation.


CHANGELOG.md
# 0.9.4 (31.05.2017
* added support for export to Excel

# 0.9.3 (13.01.2017)
* changed some bugs for rendering HTML content

# 0.9.2 (05.01.2017)
* FinancialLife now supports meta-data for account objects and payments

# 0.9.1
* minor issues with DataFrame support
* plots in examples are displayed, when examples are called from command line
* improved setup-script

# 0.9
* financial-life supports DataFrames

# 0.8
* first release on Github


FinancialLife/examples/simple_examples.py
'''
Created on 14.08.2016

@author: martin
'''
# standard libraries
from datetime import timedelta, datetime
import os

# third-party libraries
from matplotlib.pyplot import show

# own libraries
from FinancialLife.financing import accounts as a
from FinancialLife.reports import html

def example1():
    # create a private bank account and a loan
    account = a.BankAccount(amount = 1000, interest = 0.001, name = 'Main account')
    loan = a.Loan(amount = 100000, interest = 0.01, name = 'House Credit')

    # add these accounts to the simulation
    simulation = a.Simulation(account, loan)

    # describe single or regular payments between accounts. note, that
    # a string can be used for external accounts that you don't want to model.
    # also note the lambda function for the payments to the loan.
    simulation.add_regular('Income', account, 2000, interval = 'monthly')
    # you can also use lambda function to dynamically decide how much money
    # you would like to transfer
    simulation.add_regular(account, loan, lambda: min(1500, -loan.account), interval = 'monthly')

    # simulate for ten years
    simulation.simulate(delta = timedelta(days=365*10))
    # plot the data
    simulation.plt_summary()
    show()

    # print reports summarized in years
    print(account.report.yearly().as_df())
    print(loan.report.yearly().as_df())

    # analyze data
    print("Interests on bank account: %.2f" % sum(account.report.yearly().interest))
    print("Interests on loan account: %.2f" % sum(loan.report.yearly().interest))

    return simulation

def example2():
    # create a private bank account and a loan
    account = a.BankAccount(amount = 1000, interest = 0.001, name = 'Main account')
    savings = a.BankAccount(amount = 5000, interest = 0.007, name = 'Savings')
    loan = a.Loan(amount = 100000, interest = 0.01, name = 'House Credit')

    # add these accounts to the simulation
    simulation = a.Simulation(account, savings, loan)

    # describe single or regular payments between accounts. note, that
    # a string can be used for external accounts that you don't want to model.
    # also note the lambda function for the payments to the loan.
    simulation.add_regular('Income', account, 2000, interval = 'monthly')
    simulation.add_regular(account, savings, 500, interval = 'monthly')
    simulation.add_regular(account, loan, lambda: min(1500, -loan.account), interval = 'monthly')
    simulation.add_unique(savings, 'Vendor for car', 10000, '17.03.2019')

    # simulate for ten years
    simulation.simulate(delta = timedelta(days=365*10))
    # plot the data
    simulation.plt_summary()

    # print reports summarized in years
    print(account.report.yearly().as_df())
    print(loan.report.yearly().as_df())
    
    # analyze data
    print("Bank account: %.2f" % (account.account + savings.account))

    cwd = os.path.dirname(os.path.realpath(__file__))
    result_folder = cwd + '/example2'

    html.report(simulation, style="standard", output_dir = result_folder)
    show()

def example3():
    account = a.BankAccount(amount = 1000, interest = 0.001, name = 'Main account')
    savings = a.BankAccount(amount = 5000, interest = 0.013, name = 'Savings')
    loan = a.Loan(amount = 100000, interest = 0.01, name = 'House Credit')

    simulation = a.Simulation(account, savings, loan, name = 'Testsimulation')
    simulation.add_regular(from_acc = 'Income',
                           to_acc = account,
                           payment = 2000,
                           interval = 'monthly',
                           date_start = datetime(2016,9,15),
                           day = 15,
                           name = 'Income')

    simulation.add_regular(from_acc = account,
                           to_acc = savings,
                           payment = 500,
                           interval = 'monthly',
                           date_start = datetime(2016,9,30),
                           day = 30,
                           name = 'Savings')

    simulation.add_regular(from_acc = account,
                           to_acc= loan,
                           payment = 1000,
                           interval = 'monthly',
                           date_start = datetime(2016,9,15),
                           day = 15,
                           name = 'Debts',
                           fixed = False,
                           date_stop = lambda cdate: loan.is_finished())

    simulation.add_regular(from_acc = account,
                           to_acc= loan,
                           payment = lambda : min(8000, max(0,account.get_account()-4000)),
                           interval = 'yearly',
                           date_start = datetime(2016,11,20),
                           day = 20,
                           name = 'Debts',
                           fixed = False,
                           date_stop = lambda cdate: loan.is_finished())

    simulation.simulate(delta=timedelta(days=2000))

    simulation.plt_summary()
    show()

    print(account.report)
    print(loan.report)

if __name__ == '__main__':
    example1()

